go/pgprotocol/server/listener.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2025 Supabase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/multigres/multigres/go/pgprotocol/bufpool"
+)
+
+// Listener listens for incoming PostgreSQL client connections.
+type Listener struct {
+	// listener is the network listener.
+	listener net.Listener
+
+	// handler processes queries for connections.
+	handler Handler
+
+	// logger for logging.
+	logger *slog.Logger
+
+	// connBufferPooling enables buffer pooling for connections.
+	connBufferPooling bool
```

Comment:
Is there a reason we need this option? I feel like we should just choose an optimal one and live with that. I also think the default option would be to pool because it has the best of both worlds.